### PR TITLE
Add a more precise ellipsis accepted anywhere in the output

### LINKED ISSUE #40: More precises ellipsis
This kind of diff is a bit annoying:

```
-|  (cd _build/default && /home/yminsky/.opam/fresh-4.06.1/bin/ocamlc.opt -w -40 -g -bin-annot -I .freq.eobjs -I /home/yminsky/.opam/fresh-4.06.1/lib/base -I /home/yminsky/.opam/fresh-4.06.1/lib/base/caml -I /home/yminsky/.opam/fresh-4.06.1/lib/base/shadow_stdlib -I /home/yminsky/.opam/fresh-4.06.1/lib/sexplib0 -I /home/yminsky/.opam/fresh-4.06.1/lib/stdio -no-alias-deps -o .freq.eobjs/counter.cmo -c -impl counter.ml)
+|  (cd _build/default && /Users/thomas/git/rwo/book/_opam/bin/ocamlc.opt -w @a-4-29-40-41-42-44-45-48-58-59-60-40 -strict-sequence -strict-formats -short-paths -keep-locs -g -bin-annot -I .freq.eobjs -I /Users/thomas/git/rwo/book/_opam/lib/base -I /Users/thomas/git/rwo/book/_opam/lib/base/caml -I /Users/thomas/git/rwo/book/_opam/lib/base/shadow_stdlib -I /Users/thomas/git/rwo/book/_opam/lib/sexplib0 -I /Users/thomas/git/rwo/book/_opam/lib/stdio -intf-suffix .ml -no-alias-deps -opaque -o .freq.eobjs/counter.cmo -c -impl counter.ml)
```

It would be nice if `mdx` could support some way to skip some words (e.g using `<path>` or some similar keywords or notation) instead of the sometimes too generic `...` ellipsis.

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #40

### CODE REVIEW

**Comment by gpetiot** [comment]
I added unit tests for the exported functions, I'm not sure we should export more to add tests

File: lib/output.ml [hunk 1]
```diff
@@ -17,34 +17,45 @@
 open Misc
 open Compat
 
-type t = [ `Output of string | `Ellipsis ]
+type sub = S of string | Sub_ellipsis
+
+type t = [ `Output of sub list | `Ellipsis ]
```
Comment (NathanReb) [inline lib/output.ml:22]: What would you think about having a third variant for raw lines so that we can use `String.equal` for those instead of going through a regexp?

File: test/lib/test_output.ml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      let actual = of_string line in
+      Alcotest.(check (list Testable.sub)) test_name expected actual
+    in
+    (test_name, `Quick, test_fun)
+  in
+  [
+    make_test ~line:"" ~expected:[ S "" ];
+    make_test ~line:"foo" ~expected:[ S "foo" ];
+    make_test ~line:".." ~expected:[ S ".." ];
+    make_test ~line:"..." ~expected:[ S ""; Sub_ellipsis; S "" ];
```
Comment (NathanReb) [inline test/lib/test_output.ml:16]: Any particular reason why this is not simply parsing to `Sub_ellipsis`? I know it's a bit of a corner case since that will actually be recognized as an `Ellipsis` and not an `Output` but I'm more generally suprised by the leading `S ""` whenever the line starts with an ellipsis.

File: lib/output.mli [hunk 3]
```diff
@@ -16,15 +16,19 @@
...
+type t = [ `Output of sub list | `Ellipsis ]
 (** The type for test outputs. Ellipisis ([...]) allow to skip zero,
-   one or many lines while comparing the outputs with {!equal}. *)
+    one or many lines while comparing the outputs with {!equal}.
+    Sub_ellipsis ([<...>]) allow to skip zero, one or more characters or words
+    in e line, while comparing the outputs ith {!equal}. *)
 
-val equal : t list -> t list -> bool
-(** [equal x y] is true iff [x] and [y] are equivalent, modulo
+val equal : t list -> ref:t list -> bool
```
Comment (NathanReb) [inline lib/output.mli:27]: I'm a bit surprised this compares two `t list` rather than a `string list` to a `t list`, do you happen to know why?

Comment (gpetiot) [reply lib/output.mli]: it's because the arguments are swapped when calling the function recursively, I don't know how to implement it differently, so I added new functions operating on `string list` and calling the initial functions

File: lib/output.ml (hunk 1 above)
Comment (gpetiot) [reply lib/output.ml]: I added extra cases in the matches function instead

File: test/lib/test_output.ml (hunk 2 above)
Comment (gpetiot) [reply test/lib/test_output.ml]: I added more cases when parsing the string, it's fixed now

**Comment by gpetiot** [comment]
I added more tests and improved the equality tests, this is ready to be reviewed now.

**Comment by gpetiot** [comment]
@NathanReb @pitag-ha what do you think of this one?

**Comment by gpetiot** [comment]
We will work on something else that works better with the dune promote workflow instead!
